onnxruntime/core/providers/cpu/math/einsum_auxiliary_ops.cc

```diff
@@ -0,0 +1,296 @@
+        static_cast<int>(N),
+        static_cast<int>(K),
+        input_1_data + i * left_offset,
+        input_2_data + i * right_offset,
+        output_data + i * output_offset, tp);
+  }
+
+  return output;
+}
+
+template <typename T>
+Tensor ReduceSum(const Tensor& input, const std::vector<int64_t>& reduce_axes, 
+                 AllocatorPtr allocator, concurrency::ThreadPool* tp) {
+  return onnxruntime::ReduceSum<T>::Impl(input, reduce_axes, allocator, tp, true);
+}
+
+template <typename T>
+Tensor ReduceSum(const Tensor& input, int64_t axis, AllocatorPtr allocator, concurrency::ThreadPool* tp) {
+  std::vector<int64_t> reduce_axes(1, axis);
+  return ReduceSum<T>(input, reduce_axes, allocator, tp);
+}
+
+// A specific helper just for the Diagonal op
+static inline bool IsTransposeRequiredForDiagonal(int64_t dim_1, int64_t dim_2, int64_t rank) {
+  // If the input is 2D, we don't need a transpose
+  if (rank == 2)
+    return false;
+
+  // If the two dims are the innermost dims, no transpose is required
+  if ((dim_1 == rank - 1 && dim_2 == rank - 2) ||
+      (dim_1 == rank - 2 && dim_2 == rank - 1))
+    return false;
+
+  // Transpose is required
+  return true;
+}
+
+// Parse diagnoal elements along the 2 innermost dimensions
+// eg: input_shape = [1, 2, 3, 3]
+
+// This implementation provides flexibility as to which of the 2 innermost dim values is preserved 
+// via `preserve_innermost_dim_val` param
+
+// preserve_innermost_dim_val == true,
+//       output_shape = [1, 2, 1, 3] => the diagonal contains 3 elements and the dim value of the innermost dim is preserved
+
+// preserve_innermost_dim_val == false,
+//       output_shape = [1, 2, 3, 1] => the diagonal contains 3 elements and the dim value of the non-innermost dim is preserved
+static Tensor DiagonalInnermostDims(const Tensor& input, bool preserve_innermost_dim_val, AllocatorPtr allocator) {
+  const char* input_data = reinterpret_cast<const char*>(input.DataRaw());
+  const auto& input_dims = input.Shape().GetDims();
+  auto rank = input_dims.size();
+  const size_t element_size_in_bytes = input.DataType()->Size();
+
+  // This is an internal method and we already have finished all validations in the calling method.
+  // We proceed without duplicating all validations again here.
+
+  // We have a minimalistic check here to make sure the innermost dims have the same dim value
+  // as the calling method may have done a transpose before calling this method
+  ORT_ENFORCE(input_dims[rank - 2] == input_dims[rank - 1],
+              "The innermost dims should have the same dim value to parse the diagonal elements");
+
+  std::vector<int64_t> output_dims;
+  output_dims.reserve(rank);
+
+  int64_t num_iterations = 1;  // Flatten the outermost dims - this will be the number of iterations
+  for (size_t i = 0; i < rank - 2; ++i) {
+    auto input_dim_value = input_dims[i];
+    num_iterations *= input_dim_value;
+    output_dims.push_back(input_dim_value);
+  }
+
```

Comment:
Can use TensorShape::SizeToDimension and SizeFromDimension for these calcs

Reply:
But we still need to populate output_dims right ? So we can't esacpe this for loop anyway...